openedx/core/djangoapps/oauth_dispatch/models.py

```diff
@@ -71,63 +69,68 @@ class ScopedApplication(AbstractApplication):
+
     def __unicode__(self):
         """
         Return a unicode representation of this object.
         """
-        return u"<ScopedApplication '{name}'>".format(
-            name=self.name
+        return u"{application_name}:{scopes}".format(
+            application_name=self.application.name,
+            scopes=self.scopes,
         )
 
-    @property
-    def authorization_filters(self):
-        """
-        Return the list of authorization filters for this application.
-        """
-        filters = [':'.join([org.provider_type, org.short_name]) for org in self.organizations.all()]
-        if self.authorization_grant_type == self.GRANT_CLIENT_CREDENTIALS:
-            filters.append(self.FILTER_USER_ME)
-        return filters
 
-
-class ScopedApplicationOrganization(models.Model):
+class ApplicationOrganization(models.Model):
     """
-    Associates an organization to a given ScopedApplication including the
-    provider type of the organization so that organization-based filters
-    can be added to access tokens provided to the given Application.
+    Associates a DOT Application to an Organization.
 
     See openedx/core/djangoapps/oauth_dispatch/docs/decisions/0007-include-organizations-in-tokens.rst
     for the intended use of this model.
     """
-    CONTENT_PROVIDER_TYPE = 'content_org'
-    ORGANIZATION_PROVIDER_TYPES = (
-        (CONTENT_PROVIDER_TYPE, _('Content Provider')),
+    RELATION_TYPE_CONTENT_ORG = 'content_org'
+    RELATION_TYPES = (
+        (RELATION_TYPE_CONTENT_ORG, _('Content Provider')),
     )
 
-    # In practice, short_name should match the short_name of an Organization model.
-    # This is not a foreign key because the organizations app is not installed by default.
-    short_name = models.CharField(
-        max_length=255,
-        help_text=_('The short_name of an existing Organization.'),
-    )
-    provider_type = models.CharField(
+    application = models.ForeignKey(oauth2_settings.APPLICATION_MODEL, related_name='organizations')
+    organization = models.ForeignKey(Organization)
+    relation_type = models.CharField(
         max_length=32,
-        choices=ORGANIZATION_PROVIDER_TYPES,
-        default=CONTENT_PROVIDER_TYPE,
-    )
-    application = models.ForeignKey(
-        oauth2_settings.APPLICATION_MODEL,
-        related_name='organizations',
+        choices=RELATION_TYPES,
+        default=RELATION_TYPE_CONTENT_ORG,
     )
 
     class Meta:
         app_label = 'oauth_dispatch'
+        unique_together = ('application', 'organization', 'relation_type')
+
+    @classmethod
+    def get_related_org_names(cls, application, relation_type=None):
+        """
+        Return the names of the Organizations related to the given DOT Application.
+
+        Filter by relation_type if provided.
+        """
+        return [
+            r.organization.name for r in application.organizations.all()
+            if relation_type is None or r.relation_type == relation_type
+        ]
```

Comment:
Nit: It doesn't matter for small numbers, but if the number of orgs or relation types for an application is intended to grow large, flipping the composite index to be `unique_together = ('application', 'relation_type', 'organization')` would let you do this query efficiently off the index in the database.